lib/prometheus.js

```diff
@@ -0,0 +1,135 @@
+    this.pushgateway = new PromClient.Pushgateway(this.prometheusOpts.pushgatewayUrl);
+
+    debug('configure sending metrics to pushgateway');
+    this.sendMetrics(config, events)
+
+    debug('init done');
+  }
+
+  registerMetrics(prefix) {
+    this.countersStats = PromClient.register.getSingleMetric(`${prefix}_${COUNTERS_STATS}`) ||
+      new PromClient.Counter({
+        name: `${prefix}_${COUNTERS_STATS}`,
+        help: 'counter based stats e.g.: core_vusers_created_total, engine.http.requests',
+        labelNames: ['metric']
+      });
+
+    this.ratesStats = PromClient.register.getSingleMetric(`${prefix}_${RATES_STATES}`) ||
+      new PromClient.Gauge({
+        name: `${prefix}_${RATES_STATES}`,
+        help: 'rates based stats e.g.: engine_http_request_rate',
+        labelNames: ['metric'],
+      });
+
+    this.summariesStats = PromClient.register.getSingleMetric(`${prefix}_${SUMMARIES_STATES}`) ||
+      new PromClient.Gauge({
+        name: `${prefix}_${SUMMARIES_STATES}`,
+        help: 'summaries based stats e.g.: engine_http_response_time_min, engine_http_response_time_p999',
+        labelNames: ['metric'],
+      });
+
+    debug('setupMeasurements');
+  }
+
+  tagsToLabels(tags){
+    let labels = {};
+    tags.forEach(tag => {
+      let parts = tag.split(":")
+      labels[parts[0]] = parts[1]
+    })
+    return labels;
+  }
+
+  toPrometheusKey(candidate) {
+    return candidate.
+    replaceAll('.', '_').
+    replaceAll(/\s/g, '_').
+    toLowerCase();
+  }
+
+  sendMetrics(config, events) {
+    let that = this;
+
+    events.on('stats', (stats) => {
+      debug('On stats event: %O', stats);
+
+      if (stats[COUNTERS_STATS]) {
+        for (const cKey in stats[COUNTERS_STATS]) {
+          const transformed = that.toPrometheusKey(cKey)
+          this.countersStats.inc({metric: transformed}, stats[COUNTERS_STATS][cKey])
+        }
+      }
+
+      if (stats[RATES_STATES]){
+        for (const rKey in stats[RATES_STATES]) {
+          const transformed = that.toPrometheusKey(rKey)
+          this.ratesStats.set({metric: transformed}, stats[RATES_STATES][rKey])
+        }
+      }
+
+      if (stats[SUMMARIES_STATES]){
+        for (const sKey in stats[SUMMARIES_STATES]) {
+          let readings = stats[SUMMARIES_STATES][sKey];
+          for (const readingKey in readings) {
+            const transformed = `${that.toPrometheusKey(sKey)}_${readingKey}`
+            this.summariesStats.set({metric: transformed}, readings[readingKey])
+          }
+        }
+      }
+
+      // noinspection JSCheckFunctionSignatures
+      this.pushgateway.pushAdd({ jobName: this.jobUUID.toString() }, err => {
+        if (err) {
+          console.log('Error pushing metrics to push gateway', err);
+        }
+      }).then(() => {
+        debug("metrics pushed successfully")
+      });
+    });
+  }
+
+  cleanup(done) {
+    debug('cleaning up');
```

Comment:
Is there a way to ensure that all metrics have actually been sent before we exit?

Reply:
Nope, using the Pushgateway it's simple http requests, there are no guarantees.